### ISSUE TITLE
Struct with full stateful information

### ISSUE THREAD

**Issue by ShiriMoran** [np-guard/vpc-network-config-analyzer#199]
Dependency:  #255 
Currently  `func (conn *ConnectionSet) Subtract(other *ConnectionSet) *ConnectionSet` ignores the `statefullness` aspect
There are two cases here:
1. is the delta connection stateful
2. `connectionProperties` is identical but `conn stateful`  is not

The 1st item can not be computed easily since the `stateful`  property is not fine grained.
There are two ways to solve it:
1. Fine grain connection as per the stateful property; this implies that if one protocol is stateful for a certain `<src,dst>` and the other is not then they will not reside in the same connection
2. Compute the statefullness after computing the Delta for each item in the Delta

We agreed as a 1st stage to solve #170 without the stateful aspect

**Comment by adisos** [comment]
For reference:
The place where currently connections are not yet split accurately by stateful property, but rather assigned value `false` if any subset is not stateful: https://github.com/np-guard/vpc-network-config-analyzer/blob/main/pkg/vpcmodel/nodesConnectivity.go#L242-L246

The place where such more accurate split occurs (in the json output format):  https://github.com/np-guard/vpc-network-config-analyzer/blob/main/pkg/vpcmodel/jsonOutput.go#L27-L40

**Comment by adisos** [comment]
May be relevant to other operations as well (such as `union`)

**Comment by ShiriMoran** [comment]
Some connection arithmetic has an ERROR return value while other does not. This inconsistent behavior is not desired. Look into it.

**Comment by ShiriMoran** [comment]
#255 solution might result in this issue being redundant

**Comment by ShiriMoran** [comment]
*Solution* add the following struct:
```
// ExtendedSet connection details
type ExtendedSet struct {
	statefulConn    *connection.Set // stateful TCP connection between <src, dst>
	nonStatefulConn *connection.Set // nonstateful TCP connection between <src, dst>; complementary of statefulConn
	otherConn       *connection.Set // non TCP connection (for which stateful is non-relevant)
	conn            *connection.Set // entire connection
}
```
The result of connectivity analysis (across the broad) will be `ExtendedSet`; `IsStateful` will be now removed from `*Set`
`ExtendedSet` is defined as follows:
* `statefulConn`: `stateful TCP` connection 
* `otherConn: non-TCP` connection
* `conn`: the entire connection
* `nonStatefulConn`: `TCP` minus `statefulConn`
Namely, operations - union, subtract, equal, intersection - for `ExtendedSet` are defined as the operation on each `statefulConn`, `otherConn` and `conn. nonStatefulConn` Is always `TCP` minus `statefulConn`

@kyorav @zivnevo any comments (this was agreed with @adisos and @haim-kermany)

**Comment by zivnevo** [comment]
Just to make sure I understand: does the following always return `true`?
`conn.Equal(statefulConn.Union(nonStatefulConn.Union(otherConn)))`

Also, is this going to implemented in `vpc-network-config-analyzer` repo or in `models` repo?

**Comment by ShiriMoran** [comment]
> Just to make sure I understand: does the following always return `true`? `conn.Equal(statefulConn.Union(nonStatefulConn.Union(otherConn)))` (`conn` == `statefulConn` + `nonStatefulConn` + `otherConn`)
> 
Yes, you understand. This is true.

> Also, is this struct going to be implemented in `vpc-network-config-analyzer` repo or in `models` repo?

For the first PR it will be implemented in the former repo. I tend to think it should remain there - since the stateful property is connectivity analysis based, and I don't see it being used by other repos. Am I missing something?

**Comment by kyorav** [comment]
I think these definitions are rather generic, do you not think it could be useful when doing optimization for generation?

**Comment by ShiriMoran** [comment]
closed by #606
